libutils/file_lib.c

```diff
@@ -1564,6 +1588,165 @@ bool FileSparseCopy(int sd, const char *src_name,
     return retval;
 }
 
+/**
+ * Copy data jumping over areas filled by '\0' greater than blk_size, so
+ * files automatically become sparse if possible.
+ *
+ * File descriptors should already be open, the filenames #source and
+ * #destination are only for logging purposes.
+ *
+ * @NOTE Always use FileSparseClose() to close the file descriptor, to avoid
+ *       losing data.
+ */
+bool FileSparseCopy(int sd, const char *src_name,
+                    int dd, const char *dst_name,
+                    SMART_SYSCALLS_UNUSED size_t blk_size,
+                    size_t *total_bytes_written,
+                    SMART_SYSCALLS_UNUSED bool *last_write_was_a_hole)
+{
+#if !SMART_FILE_COPY_SYSCALLS_AVAILABLE
+    return FileSparseCopyShoveling(sd, src_name, dd, dst_name, blk_size,
+                                   total_bytes_written, last_write_was_a_hole);
+#else
+    size_t input_size;
+    struct stat in_sb;
+    if (fstat(sd, &in_sb) == 0)
+    {
+        input_size = in_sb.st_size;
+    }
+    else
+    {
+        Log(LOG_LEVEL_ERR, "Failed to stat() '%s': %m", src_name);
+        *total_bytes_written = 0;
+        return false;
+    }
+    bool same_dev = false;
+    struct stat out_sb;
+    if (fstat(dd, &out_sb) == 0)
+    {
+        same_dev = in_sb.st_dev == out_sb.st_dev;
+    }
+    else
+    {
+        Log(LOG_LEVEL_ERR, "Failed to stat() '%s': %m", dst_name);
+        *total_bytes_written = 0;
+        return false;
+    }
+#if HAVE_DECL_FICLONE
+    if (same_dev)
+    {
+        /* If they are on the same device, first try to use FICLONE
+         * (man:ioctl_ficlone(2)) if available because nothing can beat that.
+         */
+        if (ioctl(dd, FICLONE, sd) == 0)
+        {
+            Log(LOG_LEVEL_VERBOSE, "'%s' copied successfully to '%s'", src_name, dst_name);
+            *total_bytes_written = input_size;
+            return true;
+        }
+        else
+        {
+            Log(LOG_LEVEL_VERBOSE, "Failed to use FICLONE to copy '%s' to '%s': %m", src_name, dst_name);
+            /* let's try other things */
+            errno = 0;
+        }
+    }
+#endif  /* HAVE_DECL_FICLONE */
+    if (ftruncate(dd, input_size) != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to preset size for '%s': %m", dst_name);
+        *total_bytes_written = 0;
+        return false;
+    }
+    if (in_sb.st_blocks == 0)
+    {
+        /* Nothing to copy. */
+        Log(LOG_LEVEL_VERBOSE, "'%s' copied successfully to '%s'", src_name, dst_name);
+        *total_bytes_written = input_size;
+        return true;
+    }
+    if (fallocate(dd, FALLOC_FL_KEEP_SIZE|FALLOC_FL_PUNCH_HOLE, 0, in_sb.st_blocks * 512) != 0)
```

Comment:
```suggestion
    const int block_size = 512; // Bytes
    if (fallocate(dd, FALLOC_FL_KEEP_SIZE|FALLOC_FL_PUNCH_HOLE, 0, in_sb.st_blocks * block_size) != 0)
```

Reply:
It's not really a *block size* in the storage/file system meaning. It's just a weird unit and, actually check this:

```
              This field indicates the number of blocks allocated to the file, 512-byte units, (This may be smaller than st_size/512 when the file has holes.)

              The POSIX.1 standard notes that the unit for the st_blocks member of the stat structure is not defined by the standard.  On many  implementations it is 512 bytes; on a few
              systems, a different unit is used, such as 1024.  Furthermore, the unit may differ on a per-filesystem basis.
```
:grin:

Reply:
I'm adding a comment with the above quote explaining the `512`. I find it better than introducing a potentially misleading constant.